contracts/tools/DVP.sol

```diff
@@ -547,59 +602,70 @@ contract DVP is Ownable, ERC1820Client, IERC1400TokensRecipient, ERC1820Implemen
    *
    * @param index Index of the trade to be executed.
    */
-  function executeTrade(uint256 index) external {
+  function executeTradeWithPreimage(uint256 index, bytes32 preimage) public {
     Trade storage trade = _trades[index];
     require(trade.state == State.Pending, "Trade is not pending");
 
     if(trade.executer != address(0)) {
       require(msg.sender == trade.executer, "Trade can only be executed by executer defined at trade creation");
     }
 
+    require(block.timestamp >= trade.settlementDate, "Trade can only be executed on or after settlement date");
+
     require(_tradeisAccepted(index), "Trade has not been accepted by all token holders yet");
     
     require(_tradeisApproved(index), "Trade has not been approved by all token controllers yet");
 
-    _executeTrade(index);
+    _executeTrade(index, preimage);
   }
 
   /**
    * @dev Execute a trade in the DVP contract if possible (e.g. if tokens have been esccrowed, in case it is required).
    * @param index Index of the trade to be executed.
    */
-  function _executeTrade(uint256 index) internal {
+  function _executeTrade(uint256 index, bytes32 preimage) internal {
     Trade storage trade = _trades[index];
 
     uint256 price = _getPrice(index);
 
-    (,uint256 tokenValue1 ,,,,) = abi.decode(trade.tokenData1, (address, uint256, bytes32, Standard, bool, bool));
-    (,uint256 tokenValue2 ,,,,) = abi.decode(trade.tokenData2, (address, uint256, bytes32, Standard, bool, bool));
+    uint256 tokenValue1 = trade.userTradeData1.tokenValue;
+    uint256 tokenValue2 = trade.userTradeData2.tokenValue;
 
     if(price == tokenValue2) {
-      _transferUsersTokens(index, Holder.Holder1, tokenValue1, false);
-      _transferUsersTokens(index, Holder.Holder2, tokenValue2, false);
+      _transferUsersTokens(index, Holder.Holder1, tokenValue1, false, preimage);
+      _transferUsersTokens(index, Holder.Holder2, tokenValue2, false, preimage);
     } else {
```

Comment:
What happens in case `price != tokenValue2` and `trade.userTradeData2.tradeType == TradeType.Hold`

Reply:
If the `price != tokenValue2` then its required that the tradeType for holder2 isn't a Hold